Buildable C# code samples

cippero opened the pull request:
Hi @baanders, here are the updates to make the C# SDK sample code/solution buildable.

baanders reviewed (commented):
This is looking great!!

A few overall comments:
* For the small number of changes that resulted in a new path for a sample, we should make a corresponding PR for the docs to update the paths as needed in the relevant documents. (I left more specific file comments for these)
* When I pull this down and build it, updateTSI.cs is showing some build errors that seem to be resolved by adding the Microsoft.Azure.WebJobs.Extensions.EventHubs package to the project. If you can replicate this, we might want to add that one in.

baanders commented:
Update: I just saw you've made a doc PR here: https://github.com/MicrosoftDocs/azure-docs-pr/pull/168150
That will resolve that set of comments!

cippero commented on sdks/csharp/convertRDF.cs:
```diff
@@ -1,46 +0,0 @@
-using VDS.RDF.Ontology;
```
Sounds good, I restored a copy of this file to "sdks/csharp" for now and will add the doc change to the existing PR for the azure-docs-pr repo.

cippero commented on other/csharp/convertRDF.cs:
```diff
@@ -0,0 +1,56 @@
+using VDS.RDF.Ontology;
```
Just updated the path in the existing PR for the azure-docs-pr repo.

cippero commented:
Thanks @baanders! I just added the mentioned changes. And you're right about the Microsoft.Azure.WebJobs.Extensions.EventHubs package, I must have missed the error with updateTSI.cs. Added the package in the latest commit.

baanders reviewed (approved):
Thanks for the edits. Looks good to me!

Let's see if we can get a code review and then we'll be good to go!

soumitroauddy commented on other/csharp/convertRDF.cs:
```diff
@@ -0,0 +1,56 @@
+                    };
+
+                    // Use DTDL 'extends' for super classes 
+                    IEnumerable<OntologyClass> foundSuperClasses = owlClass.DirectSuperClasses;
```
what is the purpose of this line?

baanders replied:
Reached out to author (@khilscher) to ask

baanders replied:
Copying reply from Kevin: "An OWL graph can have parent/child classes. So to understand if an OWL class is a base class or a child class, we can look for a superclass on any given OWL class. If found, we convert these to parent + child Interfaces using DTDL extends."

The full project on which this excerpt is based can be found in this repo: (Azure-Samples/RdfToDtdlConverter)[https://github.com/azure-samples/rdftodtdlconverter/tree/main/]

vinagesh replied:
This would make a good code comment.

baanders replied:
@cippero Can we actually reword the comment slightly to avoid "we" ? (Style guide recommends avoiding it where possible)
Could do this by just removing "we" and "we can" to make it a command

cippero replied:
Will do

baanders commented on sdks/csharp/eventRoute_operations.cs:
```diff
@@ -1,38 +1,51 @@
+namespace DigitalTwins_Samples
 {
-    try
+    public class EventRouteOperationsSample
```
I'm not sure of the original reasoning. This is where the sample shows up in context in the docs: [Event route sample SDK code](https://docs.microsoft.com/azure/digital-twins/how-to-manage-routes?tabs=portal%2Csdk2%2Cportal3#event-route-sample-sdk-code)

Would you recommend this action with the SDK be shown using a public method?

soumitroauddy replied:
yes, it'd be good to make these public methods and also add a summary for each method (applies to all public methods in the solution)

/// <summary>
/// Description of method.
/// </summary>
/// <param name="ParamName"> Param description. </param>
/// <returns> Explain returned object. </returns>

baanders commented on sdks/csharp/model_operations.cs:
```diff
@@ -1,54 +1,77 @@
+            // <CreateModel>
+            // 'client' is an instance of DigitalTwinsClient
+            // Read model file into string (not part of SDK)
+            string dtdl = File.ReadAllText("MyModelFile.json");
```
@cippero  We can do this and add a comment like line 15 specifying that filename is the name of the model file .json

(Context for @soumitroauddy: Only the lines set aside with named tags (here, CreateModel from lines 14-19) are pulled from this source and displayed in the [document](https://docs.microsoft.com/azure/digital-twins/how-to-manage-model#upload-models), so trying to add some extra context for our readers who won't see the method signature.)

cippero replied:
Thanks for the feedback!

@baanders Sounds good, I'll add that comment.

baanders commented on sdks/csharp/queries.cs:
```diff
@@ -1,29 +1,45 @@
+namespace DigitalTwins_Samples
 {
-    await foreach(BasicDigitalTwin twin in result)
+    class QueriesSample
```
Which would you recommend?

Here's where the sample shows up in context in the docs: https://docs.microsoft.com/azure/digital-twins/how-to-query-graph#query-with-paging
(The class declaration actually isn't shown in the excerpt in documentation, but sharing in case it helps give an idea of which is more appropriate)

baanders commented on sdks/csharp/twin_operations_other.cs:
```diff
@@ -1,74 +1,97 @@
 
-    [JsonPropertyName("temperature")]
-    public double Temperature { get; set; }
+    class CustomDigitalTwin
```
Which would you recommend?

Here's where this sample shows up in context in the docs: https://docs.microsoft.com/azure/digital-twins/how-to-manage-twin#create-twins-using-manually-created-data

baanders reviewed (commented):
@soumitroauddy Thanks so much for the review! Left a few follow-up questions and comments.

vinagesh commented on other/csharp/convertRDF.cs:
```diff
@@ -0,0 +1,56 @@
+
+namespace DigitalTwins_Samples
+{
+    public class DigitalTwinsConvertRDFSample
```
Can we add some xml comments for this class explaining what it is doing?

vinagesh commented on sdks/csharp/twin_operations_other.cs:
```diff
@@ -1,74 +1,97 @@
-};
-await client.CreateOrReplaceDigitalTwinAsync<BasicDigitalTwin>("myTwinID", twin);
-// </TagPropertiesCsharp>
+        public async void CreateTwin(DigitalTwinsClient client)
```
Use 'public async Task' instead of 'public async void'

vinagesh commented on sdks/csharp/twin_operations_other.cs:
```diff
@@ -1,74 +1,97 @@
+            Console.WriteLine($"Temperature last updated on {response.Value.Metadata.Temperature.LastUpdatedOn}");
+            // </CreateTwin_noHelper>
+        }
+        public async void UpdateTwin(DigitalTwinsClient client)
```
Use 'public async Task' instead of 'public async void'

vinagesh commented on sdks/csharp/twin_operations_other.cs:
```diff
@@ -1,74 +1,97 @@
+            // </UpdateTwin>
+        }
+
+        public async void SetPropertyValues(DigitalTwinsClient client)
```
Use 'public async Task' instead of 'public async void'

baanders commented on sdks/csharp/twin_operations_other.cs:
```diff
@@ -1,74 +1,99 @@
 
-    [JsonPropertyName(DigitalTwinsJsonPropertyNames.DigitalTwinETag)]
-    public string ETag { get; set; }
+    internal class CustomDigitalTwin
```
@cippero  Did we ever get clarification on whether internal or public is appropriate here? I see other classes in this file are public, so wondering about the difference and confirming why these 2 are internal instead.

baanders commented on sdks/csharp/eventRoute_operations.cs:
```diff
@@ -1,38 +1,51 @@
-        await client.DeleteEventRouteAsync(routeId);
+        // ------------------ CREATE, LIST, AND DELETE EVENT ROUTE (Full sample) ---------------------
+        // <FullEventRouteSample>
+        private static async Task CreateEventRouteAsync(DigitalTwinsClient client, string routeId, DigitalTwinsEventRoute er)
```
@cippero  Should this method be public as well per the comment on line 8?

cippero replied:
Yes, I'll update it with the next commit.

baanders reviewed (commented):
A few last things!
